File: skiko/src/commonMain/kotlin/org/jetbrains/skia/ColorSpace.kt
package org.jetbrains.skia

import org.jetbrains.skia.impl.Library.Companion.staticLoad
import org.jetbrains.skia.impl.*

class ColorSpace : Managed {
    companion object {

        init {
            staticLoad()
        }

        val sRGB = ColorSpace(_nMakeSRGB(), false)
        val sRGBLinear = ColorSpace(_nMakeSRGBLinear(), false)
        val displayP3: ColorSpace = requireNotNull(makeRGB(TransferFunction.sRGB, Matrix33.displayP3ToXYZD50))

        /**
         * Creates a color space from a transfer function and a 3x3 transformation to XYZ D50.
         * Matrix33 offers various factories for obtaining a toXYZD50 matrix.
         *
         * @throws IllegalArgumentException If the transfer function is invalid.
         */
        fun makeRGB(transferFunction: TransferFunction, toXYZD50: Matrix33): ColorSpace {
            return try {
                Stats.onNativeCall()
                val ptr = interopScope {
                    _nMakeRGB(toInterop(transferFunction.asArray()), toInterop(toXYZD50.mat))
                }
                require(ptr != NullPointer) { "Invalid transfer function" }
                ColorSpace(ptr)
            } finally {
                reachabilityBarrier(transferFunction)
                reachabilityBarrier(toXYZD50)
            }
        }

    }


    /**
     * **Warning:** This method only converts between transfer functions while completely disregarding the gamut.
     * Hence, the result will not be what you expect when the gamuts of the two color spaces differ!
     */
    fun convert(toColor: ColorSpace?, color: Color4f): Color4f {
        var to = toColor
        to = to ?: sRGB
        return try {
            Color4f(withResult(FloatArray(4)) {
                    _nConvert(
                        _ptr,
                        getPtr(to),
                        color.r,
                        color.g,
                        color.b,
                        color.a,
                        it
                    )
            })
        } finally {
            reachabilityBarrier(this)
            reachabilityBarrier(to)
        }
    }

    internal constructor(ptr: NativePointer) : super(ptr, _FinalizerHolder.PTR, true)

    internal constructor(ptr: NativePointer, managed: Boolean) : super(ptr, _FinalizerHolder.PTR, managed)

    /**
     * @return  true if the color space gamma is near enough to be approximated as sRGB
     */
    val isGammaCloseToSRGB: Boolean
        get() = try {
            Stats.onNativeCall()
            _nIsGammaCloseToSRGB(_ptr)
        } finally {
            reachabilityBarrier(this)
        }

    /**
     * @return  true if the color space gamma is linear
     */
    val isGammaLinear: Boolean
        get() = try {
            Stats.onNativeCall()
            _nIsGammaLinear(_ptr)
        } finally {
            reachabilityBarrier(this)
        }

    /**
     *
     * Returns true if the color space is sRGB. Returns false otherwise.
     *
     *
     * This allows a little bit of tolerance, given that we might see small numerical error
     * in some cases: converting ICC fixed point to float, converting white point to D50,
     * rounding decisions on transfer function and matrix.
     *
     *
     * This does not consider a 2.2f exponential transfer function to be sRGB.  While these
     * functions are similar (and it is sometimes useful to consider them together), this
     * function checks for logical equality.
     */
    val isSRGB: Boolean
        get() = try {
            Stats.onNativeCall()
            _nIsSRGB(_ptr)
        } finally {
            reachabilityBarrier(this)
        }

    val transferFunction: TransferFunction
        get() = try {
            Stats.onNativeCall()
            TransferFunction(withResult(FloatArray(7)) {
                _nGetTransferFunction(_ptr, it)
            })
        } finally {
            reachabilityBarrier(this)
        }

    /** The toXYZD50 matrix that converts from this color space's gamut to XYZ adapted to D50. */
    val toXYZD50: Matrix33
        get() = try {
            Stats.onNativeCall()
            Matrix33(*withResult(FloatArray(9)) {
                _nGetToXYZD50(_ptr, it)
            })
        } finally {
            reachabilityBarrier(this)
        }

    override fun nativeEquals(other: Native?): Boolean {
        return try {
            Stats.onNativeCall()
            _nEquals(_ptr, getPtr(other))
        } finally {
            reachabilityBarrier(this)
            reachabilityBarrier(other)
        }
    }

    private object _FinalizerHolder {
        val PTR = ColorSpace_nGetFinalizer()
    }
}

@ExternalSymbolName("org_jetbrains_skia_ColorSpace__1nGetFinalizer")
private external fun ColorSpace_nGetFinalizer(): NativePointer

@ExternalSymbolName("org_jetbrains_skia_ColorSpace__nConvert")
private external fun _nConvert(fromPtr: NativePointer, toPtr: NativePointer, r: Float, g: Float, b: Float, a: Float, result: InteropPointer)

@ExternalSymbolName("org_jetbrains_skia_ColorSpace__1nMakeSRGB")
private external fun _nMakeSRGB(): NativePointer

@ExternalSymbolName("org_jetbrains_skia_ColorSpace__1nMakeSRGBLinear")
private external fun _nMakeSRGBLinear(): NativePointer

@ExternalSymbolName("org_jetbrains_skia_ColorSpace__1nMakeRGB")
private external fun _nMakeRGB(transferFunction: InteropPointer, toXYZD50: InteropPointer): NativePointer

@ExternalSymbolName("org_jetbrains_skia_ColorSpace__1nIsGammaCloseToSRGB")
private external fun _nIsGammaCloseToSRGB(ptr: NativePointer): Boolean

@ExternalSymbolName("org_jetbrains_skia_ColorSpace__1nIsGammaLinear")
private external fun _nIsGammaLinear(ptr: NativePointer): Boolean

@ExternalSymbolName("org_jetbrains_skia_ColorSpace__1nIsSRGB")
private external fun _nIsSRGB(ptr: NativePointer): Boolean

@ExternalSymbolName("org_jetbrains_skia_ColorSpace__1nGetTransferFunction")
private external fun _nGetTransferFunction(ptr: NativePointer, result: InteropPointer): NativePointer

@ExternalSymbolName("org_jetbrains_skia_ColorSpace__1nGetToXYZD50")
private external fun _nGetToXYZD50(ptr: NativePointer, result: InteropPointer): NativePointer

@ExternalSymbolName("org_jetbrains_skia_ColorSpace__1nEquals")
private external fun _nEquals(ptr: NativePointer, otherPtr: NativePointer): Boolean

File: skiko/src/nativeJsMain/cpp/ColorSpace.cc
#include <iostream>
#include "SkColorSpace.h"
#include "common.h"

static void unrefColorSpace(SkColorSpace* ptr) {
    // std::cout << "Deleting [SkColorSpace " << ptr << "]" << std::endl;
    ptr->unref();
}

SKIKO_EXPORT KNativePointer org_jetbrains_skia_ColorSpace__1nGetFinalizer() {
    return reinterpret_cast<KNativePointer>((&unrefColorSpace));
}

SKIKO_EXPORT KNativePointer org_jetbrains_skia_ColorSpace__1nMakeSRGB() {
    SkColorSpace* ptr = SkColorSpace::MakeSRGB().release();
    return reinterpret_cast<KNativePointer>(ptr);
}

SKIKO_EXPORT KNativePointer org_jetbrains_skia_ColorSpace__1nMakeSRGBLinear() {
    SkColorSpace* ptr = SkColorSpace::MakeSRGBLinear().release();
    return reinterpret_cast<KNativePointer>(ptr);
}

SKIKO_EXPORT KNativePointer org_jetbrains_skia_ColorSpace__1nMakeRGB
  (KFloat* tf, KFloat* mat) {
    skcms_TransferFunction transferFn = { tf[0], tf[1], tf[2], tf[3], tf[4], tf[5], tf[6] };
    skcms_Matrix3x3 toXYZ = {{{ mat[0], mat[1], mat[2] }, { mat[3], mat[4], mat[5] }, { mat[6], mat[7], mat[8] }}};
    SkColorSpace* ptr = SkColorSpace::MakeRGB(transferFn, toXYZ).release();
    return reinterpret_cast<KNativePointer>(ptr);
}


SKIKO_EXPORT void org_jetbrains_skia_ColorSpace__nConvert(
    KNativePointer fromPtr, KNativePointer toPtr, float r, float g, float b, float a, float* result) {
    SkColorSpace* from = reinterpret_cast<SkColorSpace*>((fromPtr));
    SkColorSpace* to = reinterpret_cast<SkColorSpace*>((toPtr));

    skcms_TransferFunction fromFn;
    from->transferFn(&fromFn);

    skcms_TransferFunction toFn;
    to->invTransferFn(&toFn);

    result[0] = skcms_TransferFunction_eval(&toFn, skcms_TransferFunction_eval(&fromFn, r));
    result[1] = skcms_TransferFunction_eval(&toFn, skcms_TransferFunction_eval(&fromFn, g));
    result[2] = skcms_TransferFunction_eval(&toFn, skcms_TransferFunction_eval(&fromFn, b));
    result[3] = skcms_TransferFunction_eval(&toFn, skcms_TransferFunction_eval(&fromFn, a));
}


SKIKO_EXPORT KBoolean org_jetbrains_skia_ColorSpace__1nIsGammaCloseToSRGB
  (KNativePointer ptr) {
    SkColorSpace* instance = reinterpret_cast<SkColorSpace*>((ptr));
    return instance->gammaCloseToSRGB();
}

SKIKO_EXPORT KBoolean org_jetbrains_skia_ColorSpace__1nIsGammaLinear
  (KNativePointer ptr) {
    SkColorSpace* instance = reinterpret_cast<SkColorSpace*>((ptr));
    return instance->gammaIsLinear();
}

SKIKO_EXPORT KBoolean org_jetbrains_skia_ColorSpace__1nIsSRGB
  (KNativePointer ptr) {
    SkColorSpace* instance = reinterpret_cast<SkColorSpace*>((ptr));
    return instance->isSRGB();
}


SKIKO_EXPORT void org_jetbrains_skia_ColorSpace__1nGetTransferFunction
  (KNativePointer ptr, KFloat* result) {
    SkColorSpace* instance = reinterpret_cast<SkColorSpace*>((ptr));
    skcms_TransferFunction tf;
    instance->transferFn(&tf);
    result[0] = tf.g;
    result[1] = tf.a;
    result[2] = tf.b;
    result[3] = tf.c;
    result[4] = tf.d;
    result[5] = tf.e;
    result[6] = tf.f;
}

SKIKO_EXPORT void org_jetbrains_skia_ColorSpace__1nGetToXYZD50
  (KNativePointer ptr, KFloat* result) {
    SkColorSpace* instance = reinterpret_cast<SkColorSpace*>((ptr));
    skcms_Matrix3x3 toXYZ;
    instance->toXYZD50(&matrix);
    result[0] = toXYZ.vals[0][0];
    result[1] = toXYZ.vals[0][1];
    result[2] = toXYZ.vals[0][2];
    result[3] = toXYZ.vals[1][0];
    result[4] = toXYZ.vals[1][1];
    result[5] = toXYZ.vals[1][2];
    result[6] = toXYZ.vals[2][0];
    result[7] = toXYZ.vals[2][1];
    result[8] = toXYZ.vals[2][2];
}


SKIKO_EXPORT KBoolean org_jetbrains_skia_ColorSpace__1nEquals
  (KNativePointer ptr, KNativePointer otherPtr) {
    SkColorSpace* instance = reinterpret_cast<SkColorSpace*>((ptr));
    SkColorSpace* other = reinterpret_cast<SkColorSpace*>((otherPtr));
    return SkColorSpace::Equals(instance, other);
}
